### ISSUE TITLE
Improve error messages

### ISSUE THREAD

**Issue by VandrerenErik** [notable/notable#1108]
- **macOS Catalina 10.15.4**:
- **Notable version: 1.8.4**:

The screenshots are self explanatory

<img width="712" alt="Screenshot 2020-04-06 at 05 54 13" src="https://user-images.githubusercontent.com/54407345/78524212-35a08480-77cb-11ea-8d0c-93648617ce09.png">

<img width="712" alt="Screenshot 2020-04-06 at 05 54 20" src="https://user-images.githubusercontent.com/54407345/78524213-36d1b180-77cb-11ea-838b-efa53a41bc64.png">

**Comment by fabiospampinato** [comment]
You should wrap your code in code blocks, otherwise the system will try to interpret it in many different ways.

**Comment by VandrerenErik** [comment]
This is not closed and you don't get the point.

I know that, the average user does NOT know that. and even if that was the case you still have a bug. The software should not render that on screen.

**Comment by fabiospampinato** [comment]
It looks like the software is trying to render some AsciiMath, as it interprets  stuff between `&&` as AsciiMath, that gets translated to TeX and then rendered by KaTeX, as that code isn't actually AsciiMath KaTeX can't render it properly and it displays an error instead.

What's the bug here? It's impossible for the parser to guess what the user actually meant, and I don't think errors should be just ignored in such cases.

People that aren't too familiar with Markdown might not know about code blocks, and that's why the app has a built-in cheatsheet. There's no other way to address this other than providing a WYSIWYG #284 editor that can't be misused like that.

**Comment by VandrerenErik** [comment]
The "bug" is an unexpected and improper behaviour from the software. In all the other markdown apps that I use in the same situation I will simply get plain text instead of rendering the error.

**Comment by VandrerenErik** [comment]
Can you parse it?

yes = do it
no = don't

But do not show the error to the user, or at least not in that way. Debug information is for developers not for users.

**Comment by fabiospampinato** [comment]
So the bug is that we have AsciiMath support and other editors don't?

That's actually an error that matters only to users and not to me, if you meant to write AsciiMath then it tells you where the error might be in what you wrote, if you meant to write plain text now you know that we support AsciiMath. If I received a thousand of those errors instead it would be pretty much worthless to me, as there's nothing here that's actually broken as far as I can tell.

Another option is just rendering plain text when an error is encountered, which I think is what you are proposing, but that's not good because if you actually meant to write AsciiMath it will be more difficult to spot the error if no error is displayed to you, and writing code as plain text like that will eventually lead to parts of it being interpreted as italic, bold or something else, it's just a matter of time so if this little error discourages you from writing code like plain text like that then I'd argue it's actually a feature.

**Comment by VandrerenErik** [comment]
No, what I'm saying is: do not show to the average user what they are going to interpret as a bug, that is the reason behind my quatation marks in bug.

As I said in my previous entry, do not show debug information to the user, at least not by default.

**Comment by fabiospampinato** [comment]
I don't agree with that sentiment 🤷‍♂️

**Comment by VandrerenErik** [comment]
That's okay, forking and do it my way is always an option.

**Comment by fabiospampinato** [comment]
Sure, however newer versions of the app aren't open-source.

Isn't it just better to wrap those lines of code in code blocks?

**Comment by VandrerenErik** [comment]
You still not get the point...

It is ugly, you don't show that to the user. For all the user care and know the app is broken by doing that.

But anyways if you are closing the sourcce code it will be better to fork it anyways ;)

**Comment by fabiospampinato** [comment]
From my point of view you're arguing for something like an IDE to not show syntax errors, which doesn't make sense.

But to each their own, it's good that we don't all think the same way.

> But anyways if you are closing the sourcce code it will be better to fork it anyways ;)

In that case you might want to put your time into something like Joplin instead: https://github.com/laurent22/joplin

**Comment by VandrerenErik** [comment]
Mate, that is exactly my point... this is NOT an IDE, I'm not developing software with this and you are literally closing the code... Why would I need debug information? it is sort of stupid actually...

**Comment by fabiospampinato** [comment]
The Markdown is kind of the source code of your notes though, it needs to be parse and compiled to HTML.

**Comment by VandrerenErik** [comment]
I told you that the problem is the way it is showed, not that is showed.

But it does not matter, I already forked it from other user that forked it when it was still open source. I already have it on my gitlab account to do it my way. So cheers

**Comment by fabiospampinato** [comment]
> I told you that the problem is the way it is showed, not that it is showed.

You said:

> The software should not render that on screen.

So where should that error be rendered then?

**Comment by VandrerenErik** [comment]
It should render something not cryptic for the average user. It should render something my grandma would understand or nothing.

But what I would personally do is give the user the options to decide what they want to see.

**Comment by VandrerenErik** [comment]
Something as basic as "Syntax error" by default will be more pleasant to the eye and will not be treated as a bug by the user

**Comment by VandrerenErik** [comment]
Okay, I like that title change

**Comment by fabiospampinato** [comment]
> It should render something not cryptic for the average user. It should render something my grandma would understand or nothing.

That I can do, depending on the computer literacy of your grandma of course, I can't quite rewrite what each third-party library (KaTeX, mermaid, plantuml etc.) writes in its own error messages, but I can wrap that in a more understandable error for sure.

> But what I would personally do is give the user the options to decide what they want to see.

Once plugins #128 are implemented you'll be able to disable AsciiMath rendering if you don't want that, and plugins will be able to implement whatever options they like.

**Comment by VandrerenErik** [comment]
We have try and catch for something. There is no need to rewrite any third party library.

**Comment by fabiospampinato** [comment]
Yeah but I mean if a third-party library throws cryptic messages it might not be possible to extract any information that could be displayed in a more user-friendly way.

**Comment by VandrerenErik** [comment]
Now... if I can make you change your mind and go back to open source and maybe follow a model like Bitwarden's... you can profit with open source more than you think. But that is your choice of course.

**Comment by fabiospampinato** [comment]
Let's keep this open as I haven't address this yet.

I've written in the subreddit and here in other issues why I don't think that's the case, most probably, but if anybody would like to take the open-source route instead I'd be very interested in watching that develop.

**Comment by VandrerenErik** [comment]
Sure.

My idea was to modify it only for my own personal use and maybe a few people in my family. I'm not really keen on the idea of creating unfedarated copies. But who knows if you keep evolving and I don't like your way I might open-source.

**Comment by fabiospampinato** [comment]
Which version did you fork, v1.5.1? (the latest one available)

**Comment by VandrerenErik** [comment]
https://github.com/zysndy/notable

**Comment by fabiospampinato** [comment]
Keep in mind that that's AGPL-licensed, not open-sourcing derivative work is a breach of the license.

**Comment by VandrerenErik** [comment]
Why would I not open-source it?

**Comment by fabiospampinato** [comment]
You said you _might_ open source it:

> But who knows if you keep evolving and I don't like your way I might open-source it.

**Comment by VandrerenErik** [comment]
Yes, because you closed the source code...

**Comment by fabiospampinato** [comment]
Whatever I do with future versions don't matter as far as the AGPL license v1.5.1 was licensed with is concerned.

**Comment by VandrerenErik** [comment]
I think you missunderstood me, if I decide to publicly fork the project it will be using the proper license and abiding it
